docs/rfc/0003-collaborative-objects.md

```diff
@@ -0,0 +1,649 @@
+layer.
+
+## Design Goals
+
+- Local first
+- Extensible. It should be possible to extend Radicle with new collaborative
+  data types without changing the protocol version
+- Interoperable. There should be a straightforward API for disparate tools to
+  interact with collaborative objects
+
+Local first because you should not need to be online to modify collaborative
+objects and no one should be able to stop you from changing data that is on
+your computer. That said, collaborative objects are _collaborative_, users need
+to be able to receive changes to the same data from other people and this
+raises the problem of how to merge those changes. I see two options here:
+
+- CRDTs, data structures which allow conflict free merges
+- Application level merges. Much as Git requires user action to merge
+  conflicting changes, we could expose the merge points of collaborative
+  objects to the user to resolve.
+
+The latter here is undesirable for a lot of reasons but most relevant is that
+it conflicts with extensibility and interoperability. In order to provide a good 
+UX, tools would need to provide a UI for users to do a three-way merge of the 
+data they are working on. This contradicts the interoperability goal.
+
+This leaves us with CRDTs. For the purposes of this RFC I am going to assume
+that we will use the [Automerge](https://github.com/automerge/automerge) CRDT
+implementation. Automerge is capable of merging arbitrary JSON objects (the
+data model is actually richer than JSON, with types for byte arrays, various
+precision integers and floats etc.). An alternative approach would be to write
+a custom CRDT for each data type we want to replicate, see [Alternative Approaches](#alternative-
+approaches) for a discussion of this design.
+
+## Example: Issues
+
+To motivate and contextualise this proposal we introduce a running example
+which is dear to many peoples hearts; issues. We imagine this from the
+perspective of the application developer. For expositional purposes we assume
+that the application communicates with a radicle implementation via an HTTP
+RPC. However, the HTTP RPC is not a proposal of this RFC, the proposed API will
+be specified in terms of `librad` later in this document.
+
+The first thing the developer must do is decide on the schema of their data and
+represent it as a JSON schema. We use this simple schema:
+
+```json
+{
+  "type": "object",
+  "rad_signed_by": {
+    "fields": ["title", "description", "author"],
+    "keys": ["<authors URN>"]
+  },
+  "properties": {
+    "id": {
+      "type": "string",
+      "frozen": true
+    },
+    "title": {
+        "type": "string",
+        "automerge_type": "text"
+    },
+    "description": {
+        "type": "string",
+        "automerge_type": "text"
+    },
+    "author": {
+        "description": "The radicle ID of the author of the issue",
+        "const": "<the authors URN>",
+        "frozen": true
+    },
+    "comments": {
+        "type": "array",
```

Comment:
It's more complex, I've reworked the example to represent comments as a tree structure with some validation that a comment is the child of a particular comment ID. This is via a `parent_id` on the comment which is then signed by the `author`. This still allows for reordering within the comments of a particular comment - but that's somewhat unavoidable in the face of concurrent modifications anyway.

Reply:
Can you explain how this prevents anybody from removing a comment from the `comments` array? Could you may be give a concrete example?

Reply:
I've added a `grow_only` key to the `children` of a comment, which should address this concern.

Reply:
Does `grow_only` also mean that existing items are immutable and does it also apply to the top-level `comments` property? If so, you can’t add nested comments because it would mutate a comment.

```
doc.get("comments).idx(0).get("children").push(...)
```
This wouldn’t work because it would mutate the first comment.

On the other hand, if the items are not immutable a comment could be completely replaced by a malicious actor. They just need to sign the replacement comment with their own signature.

Reply:
These are all good points and lead me to think that the schema based approach for validating access control is a dead end. I'm investigating an alternative capability based access control system which would be separate to the schema. I think JSON schema still makes sense for the basic structure of the data though.